src/query-client/hooks/permission.js

```diff
@@ -0,0 +1,64 @@
+import { useQuery } from 'react-query';
+import { List, Map } from 'immutable';
+import {
+  buildPermissionKey,
+  buildRolePermissionKey,
+  buildRolesPermissionsKey,
+} from '../config/keys';
+import * as Api from '../api';
+
+export default (queryClient, queryConfig) => {
+  const { retry, cacheTime, staleTime } = queryConfig;
+  const defaultOptions = {
+    retry,
+    cacheTime,
+    staleTime,
+  };
+  const useRolePermissions = (roleId) =>
+    useQuery({
+      queryKey: buildRolePermissionKey(roleId),
+      queryFn: () =>
+        Api.getRolePermissions({ id: roleId }, queryConfig).then((data) =>
+          List(data),
+        ),
+      onSuccess: async (permissions) => {
+        // save items in their own key
+        // eslint-disable-next-line no-unused-expressions
+        permissions?.forEach(async (permission) => {
+          const { id } = permission;
+          queryClient.setQueryData(buildPermissionKey(id), Map(permission));
+        });
+      },
+      ...defaultOptions,
+      enabled: Boolean(roleId),
+    });
+
+  const useRolesPermissions = (roleIds) =>
+    useQuery({
+      queryKey: buildRolesPermissionsKey(roleIds),
+      queryFn: () =>
+        Api.getRolesPermissions({ ids: roleIds }, queryConfig).then((data) => {
+          const dataImmutable = data.map((permissions) => {
+            return List(permissions);
+          });
+          return List(dataImmutable);
```

Comment:
You could use `fromJS()` it would convert the whole object to an immutable. However you would have a `List(List(Map))`.

Reply:
Yes I tried it but as you said it will give me List(List(Map)) which wouldn't work with the table